README.md
# Styled Components Breakpoints
Provides a mixin for Styled Components to handle media query widths & ranges by logical names.

## Installation
    yarn add styled-components-breakpoints
    # or
    npm install styled-components-breakpoints

## Usage

    import breakpoints from 'styled-components-breakpoints'

    const media = breakpoints({
      xxs: 0,
      xs: 320,
      s: 576,
      m: 768,
      l: 992,
      xl: 1200,
    })

If you just want to use the default breakpoints (Which are shown above):

    import breakpoints, { defaultBreakpoints }  from 'styled-components-breakpoints'

    const media = breakpoints(defaultBreakpoints)

It is recommended that the `const media` be placed in a file so that `breakpoints` only has to be defined once. Said file can then be imported across the project.

### Min-width

    media.minWidth('m')
    // Will return a media query with a min-width of 768
    // @media only screen and (min-width: 768px)

### Max-width

    media.maxWidth('m')
    // Will return a media query with a max-width of 768
    // @media only screen and (max-width: 768px)

### Only
    breakpoint.only('m')
    // Will return a range media query between "m" and the next upper breakpoint "l"
    // @media only screen and (min-width: 768px) and (max-width: 1200px)

    breakpoint.only('m', 'xl')
    // Will return a range media query between "m" and the breakpoint passed as the second argument, "xl"
    // @media only screen and (min-width: 768px) and (max-width: 1200px)

## Example
    import breakpoints, { defaultBreakpoints }  from 'styled-components-breakpoints'

    const media = breakpoints(defaultBreakpoints)

    // Styled Components
    import styled from 'styled-components'

    const Button = styled.button`
      background: white;
      font-size: 12px;

      ${media.minWidth('m')`
        font-size: 18px;
      `}
    `

In the above example a `Button` is created with a default font-size of 12px and a media query is set for medium and larger devices. This means on `xs` and `s` devices the font-size is 12px and `m` devices and up the font-size is 18px.

## Testing

    yarn install
    yarn test


RELEASES.md
# Version 1.1.1 (2022-02-21)

Update dependencies. Replace Jest with Mocha for testing.

# Version 1.0.0 (2020-08-26)

Update dependencies and convert to Yarn 2.

# Version 0.1.0

A fork from [this library](https://github.com/wearehumblebee/styled-components-breakpoint) converted to Typescript.


test/setup.js
process.on('uncaughtException', function(e) {
  console.error(e)
})

process.on('unhandledRejection', function(e) {
  console.error(e)
})


test/jest/bdd.js
// For a detailed explanation regarding each configuration property, visit:
// https://jestjs.io/docs/en/configuration.html

const { defaults } = require('jest-config')
const path = require('path')

module.exports = {
  // All imported modules in your tests should be mocked automatically
  // automock: false,

  // Stop running tests after the first failure
  // bail: false,

  // Respect "browser" field in package.json when resolving modules
  // browser: false,

  // The directory where Jest should store its cached dependency information
  // cacheDirectory: "/var/folders/s9/b5_t98vn7r707v6r42rfwjj40000gn/T/jest_dx",

  // Automatically clear mock calls and instances between every test
  clearMocks: true,

  // Indicates whether the coverage information should be collected while executing the test
  // collectCoverage: false,

  // An array of glob patterns indicating a set of files for which coverage information should be collected
  // collectCoverageFrom: null,

  // The directory where Jest should output its coverage files
  coverageDirectory: "coverage",

  // An array of regexp pattern strings used to skip coverage collection
  // coveragePathIgnorePatterns: [
  //   "/node_modules/"
  // ],

  // A list of reporter names that Jest uses when writing coverage reports
  // coverageReporters: [
  //   "json",
  //   "text",
  //   "lcov",
  //   "clover"
  // ],

  // An object that configures minimum threshold enforcement for coverage results
  // coverageThreshold: null,

  // Make calling deprecated APIs throw helpful error messages
  // errorOnDeprecated: false,

  // Force coverage collection from ignored files usin a array of glob patterns
  // forceCoverageMatch: [],

  // A path to a module which exports an async function that is triggered once before all test suites
  // globalSetup: null,

  // A path to a module which exports an async function that is triggered once after all test suites
  // globalTeardown: null,

  // A set of global variables that need to be available in all test environments
  globals: {
    "ts-jest": {
      "tsConfig": "<rootDir>/tsconfig.json"
    }
  },

  // An array of directory names to be searched recursively up from the requiring module's location
  // moduleDirectories: [
  //   "node_modules"
  // ],

  // An array of file extensions your modules use
  moduleFileExtensions: [
    ...defaults.moduleFileExtensions,
    "ts",
    "tsx"
  ],

  // A map from regular expressions to module names that allow to stub out resources with a single module
  moduleNameMapper: {
    "^src(.*)$": "<rootDir>/src$1",
    "^test(.*)$": "<rootDir>/test$1"
  },

  // An array of regexp pattern strings, matched against all module paths before considered 'visible' to the module loader
  // modulePathIgnorePatterns: [],

  // Activates notifications for test results
  // notify: false,

  // An enum that specifies notification mode. Requires { notify: true }
  // notifyMode: "always",

  // A preset that is used as a base for Jest's configuration
  preset: "ts-jest",

  // Run tests from one or more projects
  // projects: null,

  // Use this configuration option to add custom reporters to Jest
  // reporters: undefined,

  // Automatically reset mock state between every test
  // resetMocks: false,

  // Reset the module registry before running each individual test
  // resetModules: false,

  // A path to a custom resolver
  // resolver: null,

  // Automatically restore mock state between every test
  // restoreMocks: false,

  // The root directory that Jest should scan for tests and modules within
  rootDir: path.join(__dirname, '../../'),

  // A list of paths to directories that Jest should use to search for files in
  // roots: [
  //   "<rootDir>"
  // ],

  // Allows you to use a custom runner instead of Jest's default test runner
  //runner: "jest-runner-mocha",

  // The paths to modules that run some code to configure or set up the testing environment before each test
  setupFiles: [
    "raf/polyfill",
    "<rootDir>/test/setup.js"
  ],

  // The path to a module that runs some code to configure or set up the testing framework before each test
  setupFilesAfterEnv: ["<rootDir>/test/jest/setup.js"],

  // A list of paths to snapshot serializer modules Jest should use for snapshot testing
  snapshotSerializers: ["enzyme-to-json/serializer"],

  // The test environment that will be used for testing
  // testEnvironment: "<rootDir>/test/jest/environments/test.js",

  // Options that will be passed to the testEnvironment
  // testEnvironmentOptions: {},

  // Adds a location field to test results
  // testLocationInResults: false,

  // The glob patterns Jest uses to detect test files
  testMatch: [
    "<rootDir>/test/(unit)/**/*.+(js|ts|tsx)"
  ],

  // An array of regexp pattern strings that are matched against all test paths, matched tests are skipped
  testPathIgnorePatterns: [
    "/__helpers/",
    "/node_modules/"
  ],

  // The regexp pattern Jest uses to detect test files
  // testRegex: "",

  // This option allows the use of a custom results processor
  // testResultsProcessor: null,

  // This option allows use of a custom test runner
  // testRunner: "jasmine2",

  // This option sets the URL for the jsdom environment. It is reflected in properties such as location.href
  // testURL: "http://localhost",

  // Setting this value to "fake" allows the use of fake timers for functions such as "setTimeout"
  // timers: "real",

  // A map from regular expressions to paths to transformers
  transform: {
    "^.+\\.(ts|tsx)$": "ts-jest"
  },

  // An array of regexp pattern strings that are matched against all source file paths, matched files will skip transformation
  // transformIgnorePatterns: [
  //   "/node_modules/"
  // ],

  // An array of regexp pattern strings that are matched against all modules before the module loader will automatically return a mock for them
  // unmockedModulePathPatterns: undefined,

  // Indicates whether each individual test should be reported during the run
  verbose: true,

  // An array of regexp patterns that are matched against all source file paths before re-running tests in watch mode
  // watchPathIgnorePatterns: [],

  // Whether to use watchman for file crawling
  // watchman: true,
}

test/unit/index.ts
import { expect } from '../helpers'
import media, {
  defaultBreakpoints as breakpoints,
  mediaWidthRule,
  ruleTemplate,
  mediaTemplate,
  getSmallestMedia,
  getNextMedia,
  mediaRules,
} from '../../src/index'

describe('mediaWidthRule', () => {

  it('returns min-witdh for string "minWidth"', () => {
    expect(mediaWidthRule('minWidth')).to.equal('min-width')
  })

  it('returns max-witdh for string "maxWidth"', () => {
    expect(mediaWidthRule('maxWidth')).to.equal('max-width')
  })

  /*it('returns min-width when rule is not found', () => {
    expect(mediaWidthRule('unknown')).to.equal('min-width')
  })

  it('returns min-width when rule is undefined', () => {
    expect(mediaWidthRule()).to.equal('min-width')
  })*/

})

describe('ruleTemplate', () => {

  it('Returns a width rule', () => {
    expect(ruleTemplate('min-width', 600)).to.equal('(min-width: 600px)')
  })

})

describe('mediaTemplate', () => {

  it('Returns a @media query', () => {
    expect(mediaTemplate('(min-width: 600px)')).to.equal('@media only screen and (min-width: 600px)')
  })

})

describe('getSmallestMedia', () => {

  it('should get smallest media', () => {

    expect(getSmallestMedia(breakpoints)).to.equal(320)

  })

})


describe('getNextMedia', () => {

  it('finds the next breakpoint and returns it as a number', () => {

    expect(getNextMedia({
      xs: 10, s: 20, m: 30, l: 40,
    }, 20)).to.equal(30)

  })

  it('throws an error if the passed width is equal to the widest breakpoint', () => {
    
    try {
      getNextMedia({
        xs: 10, s: 20, m: 30, l: 40,
      }, 40)
    }
    catch(e) {
      expect(typeof e).to.be.a('string')
    }

  })

})


describe('mediaRules', () => {

  it('returns min-width when the rule is "minWidth"', () => {
    expect(mediaRules(breakpoints, 'm', 'minWidth')).to.equal('(min-width: 768px)')
  })

  it('returns max-width rule when the rule is "maxWidth"', () => {
    expect(mediaRules(breakpoints, 'm', 'maxWidth')).to.equal('(max-width: 768px)')
  })

  it('returns a range rule between min and max-width when both width and bound are present', () => {
    expect(mediaRules(breakpoints, 'm', 'only', 'l')).to.equal('(min-width: 768px) and (max-width: 992px)')
  })

  it('returns a range rule between max and min-width when the bound is lower than the width', () => {
    expect(mediaRules(breakpoints, 'm', 'only', 's')).to.equal('(max-width: 768px) and (min-width: 576px)')
  })

  it('returns a range rule between the width and the next minWidthper breakpoint when the rule is "only" and the bound is undefined', () => {
    expect(mediaRules(breakpoints, 'm', 'only')).to.equal('(min-width: 768px) and (max-width: 992px)')
  })

})

describe('breakpoint', () => {

  const bp = media(breakpoints)

  describe('utilities', () => {

    describe('only', () => {
      it('Implicitly returns a range media query between breakpoint "m" and the next minWidthper breakpoint', () => {
        expect(bp.only('m')`baground: red;`[1]).to.equal('@media only screen and (min-width: 768px) and (max-width: 992px)');
      })

      it('Returns a range media query between the first and second arguemt', () => {
        expect(bp.only('m', 'xl')`baground: red;`[1]).to.equal('@media only screen and (min-width: 768px) and (max-width: 1200px)');
      })

      it('Returns a range media query stating from a higher breakpoint going maxWidth', () => {
        expect(bp.only('xl', 'm')`baground: red;`[1]).to.equal('@media only screen and (max-width: 1200px) and (min-width: 768px)');
      })
    })

    describe('maxWidth', () => {
      it('Returns a max-widht media query', () => {
        expect(bp.maxWidth('m')`baground: red;`[1]).to.equal('@media only screen and (max-width: 768px)');
      })
    })

    describe('minWidth', () => {
      it('Returns a min-widht media query', () => {
        expect(bp.minWidth('m')`baground: red;`[1]).to.equal('@media only screen and (min-width: 768px)');
      })
    })

    describe('list', () => {
      it('stores an array of breakpoints in the order they where defined', () => {
        expect(bp.list).to.eql(['xxs', 'xs', 's', 'm', 'l', 'xl'])
      })
    })
  })
})


src/index.ts
import { css } from 'styled-components'

type ACC = number | null | undefined

type Rule = 'minWidth' | 'maxWidth' | 'only'

export type Breakpoints = Record<string, number>

export const defaultBreakpoints: Breakpoints = {
  xxs: 0,
  xs: 320,
  s: 576,
  m: 768,
  l: 992,
  xl: 1200,
}

// Breakpoint
export const mediaWidthRule = (rule: Rule): string => (
  `${{ minWidth: 'min', maxWidth: 'max' }[rule] || 'min'}-width`
)

export const ruleTemplate = (rule: string, width: number): string => `(${rule}: ${width}px)`;
export const mediaTemplate = (rules: string): string => (
  `@media only screen and ${rules}`
)

export const getSmallestMedia = (breakpoints: Breakpoints): number => {
  const keys: string[] = Object.keys(breakpoints)

  return (
    Number(keys.reduce((acc: ACC, key: string): number => {
      if (acc) {
        return acc > breakpoints[key] ? breakpoints[key] : acc
      }
      return breakpoints[key]
    }, undefined))
  )
}

export const getNextMedia = (breakpoints: Breakpoints, width: number): number => {
  const keys: string[] = Object.keys(breakpoints)

  const nextBreakpoint = Number(keys.reduce((acc: ACC, key: string): number => {
    const current = breakpoints[key]

    if (acc && width - current <= 0) {
      return width - acc > Math.abs(width - current) || width - acc === 0 ? current : acc
    }

    return getSmallestMedia(breakpoints)
  }, undefined))

  if (nextBreakpoint === width) {
    throw new Error(`The breakpoint of value ${nextBreakpoint} is the heighst there is, did you mean to use "[media].minWidth"?`)
  }

  return nextBreakpoint
}

export const mediaRules =
(breakpoints: Breakpoints, widthKey: string, rule: Rule, boundKey?: string) => {
  const width = breakpoints[widthKey]
  const bound = breakpoints[boundKey]
  let baseWidthRule = mediaWidthRule(rule)
  let boudWidthRule

  let baseRule = ruleTemplate(baseWidthRule, width)
  let boundRule

  if (bound && width) {
    // Get correct rule based on width relative to bound
    baseWidthRule = mediaWidthRule(bound <= width ? 'maxWidth' : 'minWidth')
    boudWidthRule = mediaWidthRule(bound <= width ? 'minWidth' : 'maxWidth')

    baseRule = ruleTemplate(mediaWidthRule(bound <= width ? 'maxWidth' : 'minWidth'), width)
    boundRule = ruleTemplate(boudWidthRule, bound)

    return [].concat([baseRule], bound ? [boundRule] : []).join(' and ')
  }

  if (!bound && rule === 'only') {
    // Get correct rule based on width relative to bound
    boudWidthRule = mediaWidthRule(bound <= width ? 'minWidth' : 'maxWidth')
    boundRule = ruleTemplate(boudWidthRule, getNextMedia(breakpoints, width))
  }

  return [].concat([baseRule], boundRule ? [boundRule] : []).join(' and ')
};

export const getMixin = (breakpoints: Breakpoints, key: string, rule: Rule = 'minWidth') => (bound?: string) => (
  (...args) => css`
      ${mediaTemplate(mediaRules(breakpoints, key, rule, bound))}{
        ${css(...args)}
      }
    `
)

export const getMediaShorthands = (breakpoints, rule) => (
  Object.keys(breakpoints).reduce((acc: Object, key: string) => ({
    ...acc,
    [key]: getMixin(breakpoints, key, rule)(),
  }), {})
)

export const getMedia = (breakpoints: Breakpoints) => {
  const mediasUp = getMediaShorthands(breakpoints, 'minWidth')
  const list = Object.keys(breakpoints)

  return ({
    ...mediasUp,
    minWidth: (widthKey: string) => getMixin(breakpoints, widthKey, 'minWidth')(),
    maxWidth: (widthKey: string) => getMixin(breakpoints, widthKey, 'maxWidth')(),
    only: (widthKey: string, boundKey?: string) => getMixin(breakpoints, widthKey, 'only')(boundKey),
    list,
  })
}

export default getMedia
